constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"fmt"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+var trueBool = true
+
+// AddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// An object can have values not described in the schema.  A blank Type could be
+	// anything, including an object, so we should treat it the same as those types.
+	// typeWithFields := sch.Type == "" || sch.Type == "object"
+	typeWithFields := sch.Type == "" || sch.Type == "object"
```

Comment:
I'd be more comfortable with top level logic that switched based on sch.Type then handled as appropriate.  This looks like several functions in one since each if block seems independent.

Reply:
I think I understand what you mean, but I don't think the blocks are quite as independent as they may seem.  The reason I say that is because we cannot rely on `Type` being defined.  It's very often left undefined by our users, or at least included inconsistently.

For example, all of these are possible in a given schema struct:

- `properties` with `type: object`
- `properties` without `type:object`
- `items` with `type: array`
- `items` without `type: array`
- `type: object` with no `properties`
- `type: array` with no `items`

I've refactored some of the function with a `switch` statement.  That section decides whether-or-not we add `x-kub...`, and is based on `Type`.  

The latter half of the function (diving into `properties`, `items`, `additionalProperties`) can't depend on the un-reliably defined `Type`, and so I've left out outside of the `switch`.

LMK your thoughts.